src/workerd/api/streams/internal.c++

```diff
@@ -78,35 +78,155 @@ private:
+    static constexpr uint64_t MAX_BUFFER_CHUNK = DEFAULT_BUFFER_CHUNK * 4;
 
     // If we know in advance how much data we'll be reading, then we can attempt to
     // optimize the loop here by setting the value specifically so we are only
-    // allocating once. But, to be safe, let's enforce an upper bound on each allocation
-    // even if we do know the total.
-    size_t amountToRead = kj::min(MAX_BUFFER_CHUNK,
-        input.tryGetLength(StreamEncoding::IDENTITY).orDefault(DEFAULT_BUFFER_CHUNK));
-
+    // allocating at most twice. But, to be safe, let's enforce an upper bound on each
+    // allocation even if we do know the total.
+    kj::Maybe<uint64_t> maybeLength = input.tryGetLength(StreamEncoding::IDENTITY);
+
+    // The amountToRead is the regular allocation size we'll use right up until we've
+    // read the number of expected bytes (if known). This number is calculated as the
+    // minimum of (limit, MAX_BUFFER_CHUNK, maybeLength or DEFAULT_BUFFER_CHUNK). In
+    // the best case scenario, this number is calculated such that we can read the
+    // entire stream in one go if the amount of data is small enough and the stream
+    // is well behaved.
+    // If the stream does report a length, once we've read that number of bytes, we'll
+    // fallback to the conservativeAllocation.
+    uint64_t amountToRead = kj::min(limit,
+                                    kj::min(MAX_BUFFER_CHUNK,
+                                            maybeLength.orDefault(DEFAULT_BUFFER_CHUNK)));
+    uint64_t conservativeAllocation = kj::min(amountToRead, DEFAULT_BUFFER_CHUNK);
+
+    // amountToRead can be zero if the stream reported a zero-length. While the stream could
+    // be lying about it's length, let's skip reading anything in this case.
     if (amountToRead > 0) {
+      auto bytes = kj::heapArray<T>(amountToRead);
+      auto pos = bytes.begin();
+      uint64_t remaining = bytes.size();
+      uint64_t writtenToBuffer = 0;
+
+      // We will keep reading into the same buffer until it is full. Once it is full,
+      // we add it to part list, allocate a new buffer, and keep reading until we get
+      // a zero-length read amount.
       for (;;) {
-        // TODO(perf): We can likely further optimize this loop by checking to see
-        // how much of the buffer was filled and using the remaining buffer space if
-        // it is not completely filled by the previous iteration. Doing so makes this
-        // loop a bit more complicated tho, so for now let's keep things simple.
-        auto bytes = kj::heapArray<T>(amountToRead);
-        size_t amount = co_await input.tryRead(bytes.begin(), 1, bytes.size());
+        // remaining here is the remaining space in the current buffer that is left
+        // to fill. pos is the current position within the current buffer where we
+        // will begin the next read.
+        KJ_DASSERT(remaining != 0);
+        uint64_t amount = co_await input.tryRead(pos, 1, remaining);
```

Comment:
You can simplify a lot here by doing `tryRead(pos, remaining, remaining)` -- set `minBytes` to the full buffer. In that case, the read will only return fewer than `remaining` if it hits EOF. So then you don't need any logic to keep track of your offset in the current buffer and try to read into it again.

Reply:
Yeah, I considered that also. My fear would be implementations of tryRead that do not properly support `minBytes`. There's definitely an argument tho that we shouldn't worry about those.

Reply:
btw, the other thing I was considering here is replacing `AllReader` entirely with a `WritableStreamSink` implementation that collects the writes. That way we could just rely on the existing optimized pumpTo implementation and all the new sink would have to worry about is combining the results when the sequence of writes complete.

Reply:
We rely on `minBytes` working correctly in many places. It should be safe to use here. I'm not aware of any implementation that handles it incorrectly (and this is something I am always careful to look at in code reviews).

Let's go ahead and use it. It will be more efficient and should make this code much simpler.